Raven.Database/Prefetching/PrefetchingBehavior.cs

```diff
@@ -1523,8 +1530,18 @@ public void AfterDelete(string key, Etag deletedEtag)
                 return;
             }
 
-            documentsToRemove.AddOrUpdate(key, s => new HashSet<Etag> { deletedEtag },
-                                          (s, set) => new HashSet<Etag>(set) { deletedEtag });
+            if (Interlocked.Increment(ref currentlyTrackedDeletedItems) >= MaxDeletedDocumentsToTrack) 
+            {                
+                var replacementOfDocumentsToRemove = new ConcurrentDictionary<string, ImmutableAppendOnlyList<Etag>>(StringComparer.InvariantCultureIgnoreCase);
+                var prevDocumentsToRemove = documentsToRemove;
+                //The reason we clean this dictionary is because we have high throughput of deletes so it is cheaper to replace the dictionary then cleaning it.
+                //Also Clear will lock the dictionary and prevent concurrent work.
+                Interlocked.Exchange(ref documentsToRemove, replacementOfDocumentsToRemove);
```

Comment:
Isn't reference update guaranteed to be atomic in CLR mem model? i.e. `documentsToRemove= replacementOfDocumentsToRemove` provided that `documentsToRemove` is defined as `volatile`?

Reply:
Then why bother using `Interlocked.Exchange`? Aren't reference updates guaranteed to be atomic in the CLR memory model?

Reply:
There are atomic, yes. But it bugs me to do it in an unsafe way.
What is worse, I'm not sure what value the GC will see in this scenario, and we want to clear it as soon as we can